File: src/lib/cpp/DSPostgreSQLDatabase.cpp
#include "DSPostgreSQLDatabase.h"
#include "DSCasts.h"

DynSoft::DSPostgreSQLDatabase::DSPostgreSQLDatabase() : DynSoft::DSDatabase() {
	SetTty(NULL);
	SetFlags(NULL);
	SetConnection(NULL);
}

DynSoft::DSPostgreSQLDatabase::DSPostgreSQLDatabase(
	const wxString &hostname,
	const wxString &database,
	const wxString &username,
	const wxString &password,
	const wxString &port
) : DynSoft::DSDatabase(
	hostname,
	database,
	username,
	password,
	port
) {
	SetTty(NULL);
	SetFlags(NULL);
	SetConnection(NULL);
}
/*
DynSoft::DSPostgreSQLDatabase::DSPostgreSQLDatabase(
	dsPostgreSQL *db
) : DynSoft::DSDatabase(
	db->getHostname(),
	db->getDatabase(),
	db->getUsername(),
	db->getPassword(),
	db->getPort()
) {
	SetTty(db->getTty());
	SetFlags(db->getFlags());
	SetConnection(db->dsConnection);
}*/

DynSoft::DSPostgreSQLDatabase::~DSPostgreSQLDatabase() {
	delete tty;
	delete flags;
}

DynSoft::DSClassInfo DynSoft::DSPostgreSQLDatabase::GetClassInfo() {
	return DynSoft::DSClassInfo(
		DS_POSTGRESQL_DATABASE_NAME,
		DS_POSTGRESQL_DATABASE_VERSION,
		DS_POSTGRESQL_DATABASE_AUTHOR
	);
}

void DynSoft::DSPostgreSQLDatabase::SetConnection(PGconn *connection) {
	this->connection = connection;
}

bool DynSoft::DSPostgreSQLDatabase::Connect() {
	const char *hostname = DSCasts::ToChar(GetHostname());
	const char *database = DSCasts::ToChar(GetDatabase());
	const char *username = DSCasts::ToChar(GetUsername());
	const char *password = DSCasts::ToChar(GetPassword());
	const char *port     = DSCasts::ToChar(GetPort());

	connection = PQsetdbLogin(
		hostname,
		port,
		GetFlags(),
		GetTty(),
		database,
		username,
		password
	);

	delete [] hostname;
	delete [] database;
	delete [] username;
	delete [] password;
	delete [] port;

	if(PQstatus(connection) == CONNECTION_BAD)
		return false;

#if defined __WXMSW__
	//PQsetClientEncoding(connection, "LATIN9");
#endif

	return true;
}

bool DynSoft::DSPostgreSQLDatabase::IsConnected() {
	return PQstatus(connection) != CONNECTION_BAD;
}

bool DynSoft::DSPostgreSQLDatabase::Reconnect() {
	if(!IsConnected()) {
		return Connect();
	}
	return true;
}

bool DynSoft::DSPostgreSQLDatabase::Disconnect() {
	if(IsConnected()) {
		PQfinish(connection);
		return true;
	}

	return false;
}

void DynSoft::DSPostgreSQLDatabase::SetFlags(char *flags) {
	this->flags = flags;
}

char *DynSoft::DSPostgreSQLDatabase::GetFlags() const {
	return flags;
}

void DynSoft::DSPostgreSQLDatabase::SetTty(char *tty) {
	this->tty = tty;
}

char *DynSoft::DSPostgreSQLDatabase::GetTty() const {
	return tty;
}

PGconn *DynSoft::DSPostgreSQLDatabase::GetConnection() const {
	return connection;
}

File: src/lib/php/DSPostgreSQLDatabase.php
<?php

require_once "DSDatabase.php";
require_once "DSClassInfo.php";

define("DS_POSTGRESQL_DATABASE_VERSION", "0.0.1");
define("DS_POSTGRESQL_DATABASE_NAME",    "DSPostgreSQLDatabase");
define("DS_POSTGRESQL_DATABASE_AUTHOR",  "Thorsten Geppert (thorsten@ext2fs.de)");

class DSPostgreSQLDatabase extends DSDatabase {

	protected $connection = null;

	public function __construct(
		$hostname = "",
		$database = "",
		$username = "",
		$password = "",
		$port     = 5432
	) {
		parent::__construct(
			$hostname,
			$database,
			$username,
			$password,
			$port
		);
	}

	public function GetClassInfo() {
		return new DSClassInfo(
			DS_POSTGRESQL_DATABASE_NAME,
			DS_POSTGRESQL_DATABASE_VERSION,
			DS_POSTGRESQL_DATABASE_AUTHOR
		);
	}

	public function Connect() {
		$this->connection = pg_connect(sprintf(
			"host=%s dbname=%s user=%s password=%s port=%s",
			$this->GetHostname(),
			$this->GetDatabase(),
			$this->GetUsername(),
			$this->GetPassword(),
			$this->GetPort()
		));

		if(!$this->connection) {
			$this->connection = null;
			return false;
		}

		return true;
	}

	public function IsConnected() {
		return pg_connection_status($this->connection) == PGSQL_CONNECTION_OK;
	}

	public function Reconnect() {
		if(!$this->IsConnected())
			return $this->Connect();
		return true;
	}

	public function Disconnect() {
		if($this->IsConnected()) {
			pg_close($this->connection);
			return true;
		}

		return false;
	}

	public function GetConnection() {
		return $this->connection;
	}

}

?>
